crates/ursa-rpc-service/src/api.rs

```diff
@@ -97,51 +108,43 @@ where
     pub store: Arc<UrsaStore<S>>,
     pub network_send: Sender<NetworkCommand>,
     pub provider_send: Sender<ProviderCommand>,
+    pending_requests: PendingRequests,
+    client: Arc<Client>,
+    origin_config: OriginConfig,
 }
 
 #[async_trait]
 impl<S> NetworkInterface for NodeNetworkInterface<S>
 where
     S: Blockstore + Store + Send + Sync + 'static,
 {
-    async fn get(&self, cid: Cid) -> Result<Option<Vec<u8>>> {
-        if !self.store.blockstore().has(&cid)? {
-            info!("Requesting block with the cid {cid:?}");
-            let (sender, receiver) = oneshot::channel();
-            let request = NetworkCommand::GetBitswap { cid, sender };
-
-            // use network sender to send command
-            self.network_send.send(request)?;
-            if let Err(e) = receiver.await? {
-                return Err(anyhow!(
-                    "The bitswap failed, please check server logs {:?}",
-                    e
-                ));
+    async fn get(&self, cid: Cid) -> Result<Vec<u8>> {
+        let new = self.check_blockstore(cid).await?;
+        let content =
+            self.store.blockstore().get(&cid)?.ok_or_else(|| {
+                anyhow!("content was fetched but could not be found in blockstore")
+            })?;
+        if new {
+            let size = content.len() as u64;
+            if let Err(e) = self.provide_cid(cid, size).await {
+                error!("{e}")
             }
         }
-        self.store.blockstore().get(&cid)
+        Ok(content)
     }
 
     async fn get_data(&self, root_cid: Cid) -> Result<Vec<(Cid, Vec<u8>)>> {
-        if !self.store.blockstore().has(&root_cid)? {
-            let (sender, receiver) = oneshot::channel();
-            let request = NetworkCommand::GetBitswap {
-                cid: root_cid,
-                sender,
-            };
-
-            // use network sender to send command
-            self.network_send.send(request)?;
-            if let Err(e) = receiver.await? {
-                return Err(anyhow!(
-                    "The bitswap failed, please check server logs {:?}",
-                    e
-                ));
+        let new = self.check_blockstore(root_cid).await?;
+        let dag = self.store.dag_traversal(&root_cid)?;
+        if new {
+            // note: this is blocking, so we need to ensure that Provider and Network Put
+            // do not take excessive time to initiate to avoid delaying response time
+            let size = dag.iter().map(|(_, c)| c.len() as u64).sum();
```

Comment:
Definitely agree, let me look into what the gateways send for headers. I wasn't sure if it was the content size or the full size of the car file including the headers etc.

For content fetched through bitswap though, it might be a little harder. I think we'd need to construct a car and calculate a total as bitswap wouldn't be sharing the car, just the blocks themselves right?

Reply:
Yeah, we can add this func to our store crate, `size(cid: CId)`

Reply:
> content size or the full size of the car file including the headers etc

I was referring to the size of the data the user wants. Whether we should include size of car headers, I'm not sure. I guess one thing to look out for is whether the origins provide size including headers, etc., then we could just provide that? Definitely some unknowns here so would be good to document afterwards so we know what our software is doing.

Reply:
Whatever the result of the research is, let's make use it is consistent with https://github.com/fleek-network/ursa/pull/221. If work in #221 needs to be updated, we can track it in a separate issue.

Reply:
Corrected the size to be the full car instead of blocks alone
- For bitswap: implemented a method on store as suggested `car_size(cid)`
- For origin: gets the length of the body and returns that to sync_content (formerly check_blockstore), which will call provide_cid with the size

https://github.com/fleek-network/ursa/pull/278/commits/ebb750854c81e7ec038c6e740b0cc8db0d4a3085